### DISCUSSION TITLE
Please update your config: extend flash life

### DISCUSSION THREAD

**Discussion by flaviut** [emporia-vue-local/esphome#115] [Announcements]
ESPHome has been storing the daily energy data to the flash memory at a 1-minute interval. This isn't great, since it will wear out the flash prematurely.

Please add the following to your config file:

```
preferences:
  # the default of 1min is far too short--flash chip is rated
  # for approx 100k writes.
  flash_write_interval: "48h"
```

**Comment by nekorevend** [comment]
Will you also change the default interval?

**Reply by flaviut** [reply]
I can't do that. It's in the ESPHome base code, and I can't change that from a component. New users will begin with copy-pasting from the updated example config.

**Comment by rananna** [comment]
I just noticed this recommendation. My emporia has been running for over a year with old firmware (500k+writes At 1/m? ) . I understand that I may not be able to upgrade the device in future if flash fails. (,although it is still upgrading but taking more time to do so. ) What are the effects of degrading flash memory on performance? Will the vue still continue to work if flash memory fails? What is the flash memory used for in the firmware? Should I replace the unit, or can flash memory be replaced?

**Reply by flaviut** [reply]
Good question!

Different parts ("pages") of the flash are completely independent in terms of degradation. Since the writes are only happening to a specific partition (or set of pages, in this case it is the NVS partition), they should have no impact on your ability to update the firmware. The code is in an entirely different partition. The code also seems to gracefully handle not being able to read the data.

This is also a bit less dire than 500k/year might suggest. The writes are spread across multiple pages, so by some napkin math, I'd expect the actual wear per year to be (1year/sync_interval)/160.

**Reply by rananna** [reply]
Thanks for this - very clear.
Another question,  in trying to bring the yaml up to date in esphome, I notice that there are now frequency and phase angle sensors added.
However, when I try to compile , I get an error like :
[frequency] is an invalid option for [0]. Please check the indentation.

The indentation is correct. What could be the error?

**Reply by rananna** [reply]
INFO ESPHome 2023.8.2
INFO Reading configuration /config/esphome/pony-panel.yaml...
WARNING GPIO12 is a Strapping PIN and should be avoided.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq.html#why-am-i-getting-a-warning-about-strapping-pins
INFO Detected timezone 'America/Toronto'
Failed config

sensor.emporia_vue: [source /config/esphome/pony-panel.yaml:105]
  platform: emporia_vue
  i2c_id: i2c_a
  phases: 
    - id: phase_a
      input: BLACK
      calibration: 0.0233
      voltage: 
        name: Phase A Voltage
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
      
      [frequency] is an invalid option for [0]. Please check the indentation.
      frequency: 
        name: Phase A Frequency
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - id: phase_b
      input: RED
      calibration: 0.022
      voltage: 
        name: ponypanel Phase B Voltage
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
  ct_clamps: 
    - phase_id: phase_a
      input: B
      power: 
        name: ponypanel Phase A Power
        id: phase_a_power
        device_class: power
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_b
      input: A
      power: 
        name: ponypanel Phase B Power
        id: phase_b_power
        device_class: power
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_b
      input: 1
      power: 
        name: ponypanel - Split 1 Power
        id: cir1
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_a
      input: 2
      power: 
        name: ponypanel - Split 2 Power
        id: cir2
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_b
      input: 3
      power: 
        name: ponypanel - Split 3 Power
        id: cir3
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_a
      input: 4
      power: 
        name: ponypanel - Split 4 Power
        id: cir4
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_b
      input: 5
      power: 
        name: ponypanel - Split 5 Power
        id: cir5
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_a
      input: 6
      power: 
        name: ponypanel - Split 6 Power
        id: cir6
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_b
      input: 7
      power: 
        name: ponypanel - Split 7 Power
        id: cir7
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_a
      input: 8
      power: 
        name: ponypanel - Split 8 Power
        id: cir8
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_a
      input: 9
      power: 
        name: ponypanel - Refrigerator Power
        id: cir9
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_b
      input: 10
      power: 
        name: ponypanel - Microwave Power
        id: cir10
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_b
      input: 11
      power: 
        name: ponypanel - Dishwasher Power
        id: cir11
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_b
      input: 12
      power: 
        name: ponypanel - Master Bath Power
        id: cir12
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_b
      input: 13
      power: 
        name: ponypanel - Patio+Pump 1 Power
        id: cir13
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_a
      input: 14
      power: 
        name: ponypanel - Patio+Pump 2 Power
        id: cir14
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
    - phase_id: phase_b
      input: 15
      power: 
        name: ponypanel - Kitchen Burners Power
        id: cir15
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
          - multiply: 2
    - phase_id: phase_b
      input: 16
      power: 
        name: ponypanel - Oven Power
        id: cir16
        filters: 
          - sliding_window_moving_average: 
              window_size: 12
              send_every: 6
          - lambda: return max(x, 0.0f);
          - multiply: 2

**Reply by rananna** [reply]
Attached the compile log above

**Reply by rananna** [reply]
Fixed it. The external components yaml had changed also, and it compiled when I brought it up to date

**Reply by rananna** [reply]
One last question (I hope). The log file now shows warnings like this:
[09:08:52][W][component:204]: Component emporia_vue.sensor took a long time for an operation (0.05 s).
[09:08:52][W][component:205]: Components should block for at most 20-30ms.
Is this normal?

**Reply by flaviut** [reply]
Normal and unfortunately pretty much unavoidable, also brought up here: https://github.com/emporia-vue-local/esphome/discussions/196

**Reply by rananna** [reply]
Thanks one final final question:
I find the frequency is slightly off 60hz, (59.8) and the phase angle is 177 degrees as apposed to 180 degrees for a purely resistive load. What can we do with these sensors? Are they information only?

**Reply by flaviut** [reply]
it's really best to make a separate topic for questions, but [yes, these sensors are known to be very imprecise](https://github.com/emporia-vue-local/esphome/discussions/190).

Phase angle has nothing to do with load factor, just with how the transformers to your house are configured. 180° means you have a split phase system.

And yeah they're pretty much useless if you live in a country with a stable power grid.

**Comment by rananna** [comment]
I will certainly post new questions as separate topics in future.
Thanks again for you emporia mod.
Great work!
